src/backend/distributed/utils/acquire_lock.c

```diff
@@ -0,0 +1,311 @@
+{
+	BackgroundWorkerHandle *handle = (BackgroundWorkerHandle *) arg;
+	TerminateBackgroundWorker(handle);
+}
+
+
+/*
+ * Signal handler for SIGTERM
+ *		Set a flag to let the main loop to terminate, and set our latch to wake
+ *		it up.
+ */
+static void
+lock_acquire_helper_sigterm(SIGNAL_ARGS)
+{
+	int save_errno = errno;
+
+	got_sigterm = true;
+	SetLatch(MyLatch);
+
+	errno = save_errno;
+}
+
+
+/*
+ * ShouldAcquireLock tests if our backend should still proceed with acquiring the lock,
+ * and thus keep terminating conflicting backends. This function returns true until a
+ * SIGTERM, background worker termination signal, has been received.
+ *
+ * The function blocks for at most sleepms when called. During operation without being
+ * terminated this is the time between invocations to the backend termination logic.
+ */
+static bool
+ShouldAcquireLock(long sleepms)
+{
+	int rc;
+
+	rc = WaitLatch(MyLatch, WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH,
+				   sleepms * 1L, PG_WAIT_EXTENSION);
+	ResetLatch(MyLatch);
+
+	/* emergency bailout if postmaster has died */
+	if (rc & WL_POSTMASTER_DEATH)
+	{
+		proc_exit(1);
+	}
+
+	CHECK_FOR_INTERRUPTS();
+
+	return !got_sigterm;
+}
+
+
+/*
+ * LockAcquireHelperMain runs in a dynamic background worker to help master_update_node to
+ * acquire its locks.
+ */
+void
+LockAcquireHelperMain(Datum main_arg)
+{
+	int backendPid = DatumGetInt32(main_arg);
+	StringInfoData sql;
+	LockAcquireHelperArgs *args = (LockAcquireHelperArgs *) MyBgworkerEntry->bgw_extra;
+	long timeout = 0;
+	const TimestampTz connectionStart = GetCurrentTimestamp();
+	const TimestampTz deadline = TimestampTzPlusMilliseconds(connectionStart,
+															 args->lock_cooldown);
+
+	pqsignal(SIGTERM, lock_acquire_helper_sigterm);
+
+	BackgroundWorkerUnblockSignals();
+
+	elog(LOG, "lock acquiring backend started for backend %d (cooldown %dms)", backendPid,
+		 args->lock_cooldown);
+
+	/*
+	 * this loop waits till the deadline is reached (eg. lock_cooldown has passed) OR we
+	 * no longer need to acquire the lock due to the termination of this backend.
+	 * Only after the timeout the code will continue with the section that will acquire
+	 * the lock.
+	 */
+	do {
+		timeout = DeadlineTimestampTzToTimeout(deadline);
+	} while (timeout > 0 && ShouldAcquireLock(timeout));
+
+	/* connecting to the database */
+	BackgroundWorkerInitializeConnectionByOid(args->DatabaseId, InvalidOid, 0);
+
+	/* TODO explain sql below */
+	initStringInfo(&sql);
+	appendStringInfo(&sql,
+					 "SELECT \n"
+					 "    DISTINCT conflicting.pid,\n"
+					 "    pg_terminate_backend(conflicting.pid)\n"
+					 "  FROM pg_locks AS blocked\n"
+					 "       JOIN pg_locks AS conflicting\n"
+					 "         ON (conflicting.database = blocked.database\n"
+					 "             AND conflicting.objid = blocked.objid)\n"
+					 " WHERE conflicting.granted = true\n"
+					 "   AND blocked.granted = false\n"
+					 "   AND blocked.pid = %d;",
```

Comment:
Probably more hassle than it's worth, but this seems like a candidate for `SPI_execute_with_args`.

Reply:
`with_args` makes a lot of sense, will rewrite it in such a way. Even though I think it will be safe to interpolate an int in there, it always feels wrong to munge sql with string interpolation.